pkg/gateway/gateway.go

```diff
@@ -0,0 +1,122 @@
+package gateway
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"io"
+	"net"
+	"net/url"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/recover"
+)
+
+var syntheticError = errors.New("synthetic error")
+
+type Runnable interface {
+	Run(context.Context) error
+}
+
+type gateway struct {
+	env       env.Core
+	log       *logrus.Entry
+	l         net.Listener
+	allowList map[string]struct{}
+}
+
+func NewGateway(env env.Core, log *logrus.Entry, l net.Listener, acrResourceID, gatewayDomains string) (Runnable, error) {
```

Comment:
I understand this is dead simple (I would still like bit more deep dive). But did we checked azure gateway as an option for this? I know its big, clumsy, and slow. But just wondering. 

And this for this we would be good to have some limiting but not sure how to achieve this.
Or how to do load based loadbalancing once we scale the VMSS.

Reply:
I don't believe there's an alternative Azure resource that we can obviously use here.